src/nnet3/nnet-simple-component.cc

```diff
@@ -1011,6 +1022,196 @@ void PerElementScaleComponent::UnVectorize(
+  offsets_.AddVec(alpha, other->offsets_);
+}
+
+PerElementOffsetComponent::PerElementOffsetComponent(
+    const PerElementOffsetComponent &component):
+    UpdatableComponent(component),
+    offsets_(component.offsets_) { }
+
+void PerElementOffsetComponent::SetZero(bool treat_as_gradient) {
+  if (treat_as_gradient) {
+    SetLearningRate(1.0);
+    is_gradient_ = true;
+  }
+  offsets_.SetZero();
+}
+
+void PerElementOffsetComponent::PerturbParams(BaseFloat stddev) {
+  CuVector<BaseFloat> temp_offsets(offsets_);
+  temp_offsets.SetRandn();
+  offsets_.AddVec(stddev, temp_offsets);
+}
+
+std::string PerElementOffsetComponent::Info() const {
+  std::stringstream stream;
+  BaseFloat offsets_mean = offsets_.Sum() / offsets_.Dim();
+  BaseFloat offsets_stddev = std::sqrt(VecVec(offsets_, offsets_) /
+                              offsets_.Dim()) - (offsets_mean * offsets_mean);
+  stream << Type() << ", input-dim=" << InputDim()
+         << ", output-dim=" << OutputDim()
+		 << ", offsets-mean=" << offsets_mean
+         << ", offsets-stddev=" << offsets_stddev
+         << ", learning-rate=" << LearningRate();
+  return stream.str();
+}
+
+Component* PerElementOffsetComponent::Copy() const {
+  PerElementOffsetComponent *ans = new PerElementOffsetComponent();
+  ans->learning_rate_ = learning_rate_;
+  ans->offsets_ = offsets_;
+  ans->is_gradient_ = is_gradient_;
+  return ans;
+}
+
+BaseFloat PerElementOffsetComponent::DotProduct(
+    const UpdatableComponent &other_in) const {
+  const PerElementOffsetComponent *other =
+      dynamic_cast<const PerElementOffsetComponent*>(&other_in);
+  return VecVec(offsets_, other->offsets_);
+}
+
+void PerElementOffsetComponent::Init(
+    BaseFloat learning_rate, int32 dim,
+    BaseFloat param_mean, BaseFloat param_stddev) {
+  UpdatableComponent::Init(learning_rate);
+  offsets_.Resize(dim);
+  KALDI_ASSERT(dim > 0 && param_stddev >= 0.0);
+  offsets_.SetRandn();
+  offsets_.Scale(param_stddev);
+  offsets_.Add(param_mean);
+}
+
+void PerElementOffsetComponent::Init(BaseFloat learning_rate,
+                                    std::string matrix_filename) {
+  UpdatableComponent::Init(learning_rate);
+  CuMatrix<BaseFloat> mat;
+  ReadKaldiObject(matrix_filename, &mat); // will abort on failure.
+  KALDI_ASSERT(mat.NumCols() == 1);
+  int32 dim = mat.NumRows();
+  offsets_.Resize(dim);
+  offsets_.CopyColFromMat(mat, 0);
+}
+
+void PerElementOffsetComponent::InitFromConfig(ConfigLine *cfl) {
+  bool ok = true;
+  BaseFloat learning_rate = learning_rate_;
+  std::string matrix_filename;
+  int32 dim = -1;
+  cfl->GetValue("learning-rate", &learning_rate); // optional.
+  if (cfl->GetValue("matrix", &matrix_filename)) {
+    Init(learning_rate, matrix_filename);
+    if (cfl->GetValue("dim", &dim))
+      KALDI_ASSERT(dim == InputDim() &&
+                   "input-dim mismatch vs. matrix.");
+  } else {
+    ok = ok && cfl->GetValue("dim", &dim);
+    BaseFloat param_mean = 1.0, param_stddev = 0.0;
```

Comment:
Actually, I'd make the defaults zero mean and default stddev...
the reason to make stddevs nonzero is usually to ensure nonzero gradients
in multilayer networks, but that's not an issue for this type of component.
Dan


On Sun, Nov 1, 2015 at 5:11 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/307#discussion_r43588439>:
>
> > +}
> > +
> > +void PerElementOffsetComponent::InitFromConfig(ConfigLine *cfl) {
> > +  bool ok = true;
> > +  BaseFloat learning_rate = learning_rate_;
> > +  std::string matrix_filename;
> > +  int32 dim = -1;
> > +  cfl->GetValue("learning-rate", &learning_rate); // optional.
> > +  if (cfl->GetValue("matrix", &matrix_filename)) {
> > +    Init(learning_rate, matrix_filename);
> > +    if (cfl->GetValue("dim", &dim))
> > +      KALDI_ASSERT(dim == InputDim() &&
> > +                   "input-dim mismatch vs. matrix.");
> > +  } else {
> > +    ok = ok && cfl->GetValue("dim", &dim);
> > +    BaseFloat param_mean = 1.0, param_stddev = 0.0;
>
> I think the default initializers should be zero mean and unit std-dev
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/307/files#r43588439>.
>